lib/fswatcher/fswatcher.go

```diff
@@ -0,0 +1,558 @@
+		}
+
+		// If there are no events below path, but we need to recurse
+		// into that path, create eventDir at path.
+		if _, ok := parentDir.dirs[currPath]; !ok {
+			l.Debugf("%v Creating eventDir: %s", w, currPath)
+			parentDir.dirs[currPath] = newEventDir(currPath, parentDir)
+		}
+		parentDir = parentDir.dirs[currPath]
+
+		// Reset allowed children count to maxFilesPerDir for non-root
+		if i == 0 {
+			localMaxFilesPerDir = maxFilesPerDir
+		}
+	}
+
+	if event, ok := parentDir.events[evPath]; ok {
+		event.lastModTime = evTime
+		event.evType |= evType
+		l.Debugf("%v Already tracked (type %v): %s", w, event.evType, evPath)
+		return
+	}
+
+	childDir, ok := parentDir.dirs[evPath]
+
+	// If a dir existed at path, it would be removed from dirs, thus
+	// childCount would not increase.
+	if !ok && parentDir.childCount() == localMaxFilesPerDir {
+		l.Debugf("%v Parent dir already has %d children, tracking it instead: %s", w, localMaxFilesPerDir, evPath)
+		w.aggregateEvent(filepath.Dir(evPath), evTime, evType)
+		return
+	}
+
+	firstModTime := evTime
+	if ok {
+		firstModTime = childDir.getFirstModTime()
+		evType |= childDir.getEventType()
+		delete(parentDir.dirs, evPath)
+	}
+	l.Debugf("%v Tracking (type %v): %s", w, evType, evPath)
+	parentDir.events[evPath] = &event{
+		path:         evPath,
+		firstModTime: firstModTime,
+		lastModTime:  evTime,
+		evType:       evType,
+	}
+	w.resetNotifyTimerIfNeeded()
+}
+
+func (w *watcher) actOnTimer() {
+	eventCount := w.rootEventDir.eventCount()
+	if eventCount == 0 {
+		l.Debugln(w, "No tracked events, waiting for new event.")
+		w.notifyTimerNeedsReset = true
+		return
+	}
+	oldFsEvents := w.popOldEvents(w.rootEventDir, time.Now())
+	if len(oldFsEvents) == 0 {
+		l.Debugln(w, "No old fs events")
+		w.resetNotifyTimer(w.notifyDelay)
+		return
+	}
+	// Sending to channel might block for a long time, but we need to keep
+	// reading from notify backend channel to avoid overflow
+	go func() {
+		timeBeforeSending := time.Now()
+		l.Debugf("%v Notifying about %d fs events", w, len(oldFsEvents))
+		separatedBatches := make(map[eventType][]string)
+		for path, event := range oldFsEvents {
+			separatedBatches[event.evType] = append(separatedBatches[event.evType], path)
+		}
+		for _, evType := range [3]eventType{nonRemove, mixed, remove} {
+			if len(separatedBatches[evType]) != 0 {
+				select {
+				case w.notifyChan <- separatedBatches[evType]:
+				case <-w.stop:
+					return
+				}
+			}
+		}
+		// If sending to channel blocked for a long time,
+		// shorten next notifyDelay accordingly.
+		duration := time.Since(timeBeforeSending)
+		buffer := time.Duration(1) * time.Millisecond
+		var nextDelay time.Duration
+		switch {
+		case duration < w.notifyDelay/10:
+			nextDelay = w.notifyDelay
+		case duration+buffer > w.notifyDelay:
+			nextDelay = buffer
+		default:
+			nextDelay = w.notifyDelay - duration
+		}
+		select {
+		case w.notifyTimerResetChan <- nextDelay:
+		case <-w.stop:
+		}
+	}()
```

Comment:
Thanks for the reminder, forgot about this.
About your first comment: I don't understand how this can spawn multiple times. The timer triggering a scan schedule is only reset after it finishes.

About the cancellation: This is run as a suture service controlled by the model, it is thus stop at the same time as the versioners and  folder (actually just after versioners and before the folder). Controlling this by the folder context and suture service feels redundant.